joomla/joomla-cms
# More flexible requires attribute in form list options (+Extend to form fields)

Pull Request for Issue https://github.com/joomla/joomla-cms/issues/12517.
### Summary of Changes

Add more flexibility to `requires` attribute in form list options.
Also, extend funcionality of the requires to the main fields. Do not setup the fields if requirements are not fulfilled.
### Testing Instructions
- Do a code review
- Apply this patch on 3.7.x
- Add this to a joomla form xml

``` xml
<field
    name="test 1"
    type="list"
    label="Test 1"
    description="Test 1"
    >
    <option value="">Check</option>
    
    <option value="a1" requires="multilanguage">B/C multilanguage</option>
    <option value="a2" requires="associations">B/C associations</option>
    <option value="a3" requires="vote">B/C vote</option>
    
    <option value="b1" requires="com_search">com_search</option>
    <option value="b2" requires="com_search{show_date:1}">com_search{show_date:1}</option>
    <option value="b3" requires="config{ftp_enable:1}">config{ftp_enable:1}</option>
    <option value="b4" requires="config{mailonline:1[AND]mailer:smtp}">config{mailonline:1[AND]mailer:smtp}</option>
    <option value="b5" requires="config{mailonline:1[AND]mailer:smtp},com_search{show_date:1}">config{mailonline:1[AND]mailer:smtp},com_search{show_date:1}</option>
    <option value="b6" requires="plg_system_p3p">plg_system_p3p</option>
    <option value="b7" requires="plg_system_languagefilter{item_associations:1}">plg_system_languagefilter{item_associations:1}</option>
</field>

<field
    name="test2"
    type="text"
    label="Test 2"
    description="Test 2"
    hint="Test 2"
    requires="config{mailonline:1[AND]mailer:smtp}"
/>

<field
    name="test3"
    type="integer"
    label="Test 3"
    description="Test 3"
    first="1"
    last="10"
    step="1"
    requires="config{mailonline:1[AND]mailer:smtp},com_search{show_date:1}"
/>
```
- Check that the options in this list field only are displayed when the requirements are fulfilled.
- Check the second and third field only exists when the requirements are fulfilled.
- Make other tests with other requires you can remember.
### Documentation Changes Required

Maybe add this to the form fields documentations.
### Notes

If merged i will change all existent require attributes to this new form.

@mbabker please review/test.
